OpenTelemetry Sandbox

jpkrohling opened the pull request:
Closes https://github.com/open-telemetry/oteps/issues/231


Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

yurishkuro commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+
+There is a desire, but not an expectation, that projects will be moved from the sandbox as an official SIG or incorporated into an existing SIG. There’s also no expectation that the OpenTelemetry project will provide resources to the sandbox project, like extra GitHub CI minutes or Zoom meeting rooms, although we might evaluate individual requests.
+
+Projects in the sandbox should leave it within one year, either by being merged into existing SIGs, being accepted as independent SIGs, or by being archived.
```
this directly contradicts the previous paragraph. Even CNCF does not have this "leave in one year"(but it does have annual reviews).

jpkrohling replied:
What I'm trying to convey is that it's OK to experiment and to have failed experiments. So, there's a hope, but not expectation that they will "graduate". I added the following, do you think it's closer to what I mean?

> Realistically, we know that experiments might get dropped.

yurishkuro replied:
I think what you want is an annual review with possible outcomes being:
  * continue as sandbox for another year
  * graduate to official OTEL
  * abandon project and archive the repo
  * transition to another org if there is a better fit

This is not the same as "should leave within one year", which is often unrealistic, nor has a strong justification.

jpkrohling replied:
The concern I had was in having projects in the sandbox indefinitely, without giving enough motivation to move forward. Annual reviews would solve this, as the project might be folded after one year when there's not enough activity.

jpkrohling replied:
If code is made as part of this organization and is open-source, should we have an official "transition" step here, or just assume that people will fork if they want to continue working on that? Would "transitioning to another org" mean that we'd legally transfer the assets there (name, copyright of the code)? If the project is named "opentelemetry-data-visualizer", are we OK in having another entity own it with this name?

I would keep this off for now, and just let people fork the archived repository.

yurishkuro commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+A low barrier to entry would be desired for the sandbox. While the process can be refined based on our experience, my initial proposal for the process is the following:
+
+1. Proposals should be written following the template below and have one Technical Committee (TC) and/or Governance Committee (GC) sponsor, who will regularly provide the TC and GC information about the state of the project.
+1. Once a sponsor is found, the TC and GC will vote on accepting this new project on the Slack channel #opentelemetry-gc-tc. 
```
This is an acceptance process, not acceptance criteria. On what grounds are TC/GC expected to decide?

jpkrohling replied:
I would leave this open for now. Each voting member can accept/reject the proposal based on their best judgement on what's best for the project.

yurishkuro commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+> 
+> Repository name:
+>  
+> Motivation:
```
Motivation for the project / product or for joining the sandbox? I would require both.

In general, I think this should be structured to impost as little as necessary overhead on GC/TC and as much as possible on the project authors. If we have specific acceptance criteria / checklist, it should be filled out in the project proposal. See CNCF process - there is an extensive checklist, including for sandbox projects.

jpkrohling replied:
> Motivation for the project / product or for joining the sandbox? I would require both.

Good idea. I had a "use-case" section before, but removed when I added the example and felt that stating the use-case in the motivation would be enough.

yurishkuro commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+
+### Periodic reports
+
+On the first [GC/TC joint call](https://docs.google.com/document/d/1jylE5uZCKV9mrPw8Qrc5ExGyRVbBdqcbWPwni-hB5dE) of the calendar quarter (Jan, Apr, Jul, Oct), the TC/GC sponsors for the sandbox projects MUST provide an update about their sponsored project. Should the sponsor not be able to join the call, a short written report MUST be provided as part of the meeting notes.
```
👎

jpkrohling replied:
Is this a no for the periodic reports as a whole? Or for specific parts of it? Do you have any suggestions on how to keep track of the health of the projects? My concern is that projects might get accepted and forgotten there until it's time to pull the plug. Perhaps they need a nudge from time to time?

Note that I didn't state that sponsors should write the report themselves. They are on the hook for providing those reports: they can be written by sandbox members and validated by the sponsor before being relayed to the bigger group.

yurishkuro replied:
I answered above https://github.com/open-telemetry/oteps/pull/233/files#r1224666071

jpkrohling replied:
I'm marking this as resolved, as the discussion is happening on the issue you linked in the previous comment.

jpkrohling commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+
+A low barrier to entry would be desired for the sandbox. While the process can be refined based on our experience, my initial proposal for the process is the following:
+
+1. Proposals should be written following the template below and have one Technical Committee (TC) and/or Governance Committee (GC) sponsor, who will regularly provide the TC and GC information about the state of the project.
```
> two people is not a low barrier, what is the motivation?

Combined, the TC and GC group has 16 members if I'm counting right, but you are right. We can have a lower number now and increase if needed. Originally, I didn't have sponsors being on the hook for providing updates, so, a simple "that's nice" would be sufficient from two sponsors. But now that there's an expectation that the sponsor will provide quarterly feedback, the barrier is already higher than 2 simple "that's nice". 

> "who will regularly provide" - hard no from me

I'll remove this from here, as I define the frequency elsewhere. Is once a quarter acceptable?

yurishkuro replied:
I am opposed to _any_ framework where it is GC/TC responsibility. When OTEL prepares for annual review with CNCF, it's on OTEL project to prepare an update, not on CNCF TOC member. But there is a well-define questionnaire that a project is supposed to answer in the review.

jpkrohling replied:
It's not the GC or TC responsibility to produce the report. The idea is that the TC/GC sponsor will provide an update about the sandbox. What I have in mind is mostly a health check, stating whether the experiment is still on track or if people just stopped working on that and we should pull the plug. This is not an update from us to the general public. Note that the TC/GC sponsor is someone who cares enough for the sandbox project, joining their regular calls. Producing this health-check report shouldn't be too much effort, especially if it's once a quarter.

yurishkuro commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+
+## Prior art and alternatives
+
+* One obvious alternative is to let users collaborate on their own accounts and organizations, as some are doing today. There are two problems with that: the first is that those projects usually lack clear governance rules, so that external contributors aren't sure what's going to happen with their code contributions. The second problem is lack of visibility: most of the current experiments and i ...
```
> those projects usually lack clear governance rules

Nothing prevents those projects from adopting the same license and code of conduct as CNCF/OTEL.

> lack of visibility

This proposal in no way addresses the visibility problem. If the attempt is to address is by simply being in the otel-sandbox list of repos (which isn't very search friendly), then there are much easier ways of achieving the same (e.g. a page / registry on the website with minimum entry requirements) than having a sandbox program.

jpkrohling replied:
Being part of the OTel repository certainly makes it more visible: a blog post mentioning "github.com/open-telemetry-sandbox/some-experiment" would certainly get more attention than "github.com/jpkrohling/some-experiment". Other than that, you are right that there are no active steps being taken to make those repos more visible.

yurishkuro commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+
+As a possible solution to this, this OTEP proposes a new GitHub organization, [opentelemetry-sandbox](https://github.com/opentelemetry-sandbox). This organization will host projects until there's confidence that they have a healthy community behind them. They would also serve as a neutral place for the community to conduct experiments.
+
+The advantage of a sandbox organization is that OpenTelemetry's governance rules can be applied, making sure it’s an inclusive place for people to collaborate while keeping the reputation of the OpenTelemetry project as a whole untouched, given that it would be clear that OpenTelemetry doesn’t officially support projects within the sandbox.
```
> governance rules can be applied

What specifically does that mean? That TC is expected to arbitrate technical disagreements? Even that does not remove the need for internal governance of the proposed project.

jpkrohling replied:
I had CoC violations in mind when I wrote this but didn't want to limit the scope to that, as I can't foresee the problems we'll have.

> That TC is expected to arbitrate technical disagreements?

Great question, I'll add this to the open questions.

jpkrohling commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,89 @@
+## Prior art and alternatives
+
+* One obvious alternative is to let users collaborate on their own accounts and organizations, as some are doing today. There are two problems with that: the first is that those projects usually lack clear governance rules, so that external contributors aren't sure what's going to happen with their code contributions. The second problem is lack of visibility: most of the current experiments and i ...
+* Another alternative is to direct users to CNCF's sandbox, but at this moment, the queue of projects applying for that is huge and there's no expectations that a project applying today would be considered within the next few months. Besides, the CNCF Sandbox implies that projects are standalone projects, while OpenTelemetry Sandbox projects would often be incorporated as part of an existing SIG.
```
I like your optimism on the success of the sandbox. If we end up having such a big number of projects applying to the sandbox, I'm sure we can figure out a solution.

jpkrohling replied:
How about a compromise: we commit to having at most 3 projects in the sandbox for the first year. It shouldn't have too much overhead for the TC and we can use this experience to see if this would scale.

jpkrohling replied:
I added this limit of 3 projects to the proposal.

MSNev commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,93 @@
+
+### Exiting the sandbox
+
+Projects can be hosted for at most 12 months and are expected to exit before that, with one of the following outcomes:
```
12 Months seems a bit of a hard limit, perhaps rephrasing that once the sandbox project has not been "active" for a certain period of time.

MSNev replied:
For example there is already a sandbox for Web JS exprementation https://github.com/open-telemetry/opentelemetry-sandbox-web-js, and having a general 12 month limit would not work for something like that project.

It's also a little different in that we have automated scripts for merging from the "main" repo's into the sandbox (with all commit history) to avoid creating disconnected forks and explicitly stating that the all successful experiments from the sandbox MUST be contributed back to the main repo's.

Part of this is that we explicitly prepend `sandbox-` to every package, and while there are currently no plans for releasing packages from this sandbox, by having the prefix it more explicitly calls out the non-production nature of anything produced from this repo and any branches that it will have.

jpkrohling replied:
I'm fine removing this part altogether. I added this based on the feedback from @pyohannes: https://github.com/open-telemetry/oteps/issues/231#issuecomment-1580257623

I like forcing a decision after one year, as it would either show that a community could be formed around the subject of the experiment and will continue working on that once it becomes part of a SIG, or that the experiment completed with a concrete outcome (success, failure). I'm not sure about the web-js sandbox, but I'd be interested in knowing more about its timeline and current state.

jpkrohling commented:
@yurishkuro, I think I addressed most (all?) your comments. Let me know if they are acceptable, in which case I'd like to discuss the open question we have there about the arbitration of technical disagreements.

jpkrohling commented:
@yurishkuro, can you take another look at this one? I believe I addressed your concerns, but I'd be interested in hearing if you'd still be against the general idea. 

I'm also interested in hearing from others, including the wider community, maintainers, and TC members.

yurishkuro reviewed (commented):
I don't think my or other people's concerns are addressed. You wrote some comments in the discussion, but the OTEP text still contains the same burdensome rules that impose an unnecessary process that takes time away from GC/TC, like frequent reviews / updates.

jpkrohling commented:
I updated this PR with a few changes, addressing the concerns expressed. I would appreciate another review round.

svrnm commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,92 @@
+    3. The voting closes automatically once a simple majority of the TC/GC electorate has chosen one side.
+3. Proponents should abide by OpenTelemetry’s Code of Conduct (currently the same as CNCF’s).
+4. There’s no expectation that small sandbox projects will have regular calls, but there is an expectation that all decisions will be made in public and transparently.
+5. Sandbox projects do NOT have the right to feature OpenTelemetry’s name on their websites.
```
Can you explain the rational behind this? I am not against the statement, I am more worried about the "on their websites", why should a sandbox have a website at all?

jpkrohling replied:
Perhaps this is me trying to anticipate the resolution of hypothetical problems, but I want to avoid something getting into the sandbox and then claiming all over the world that it is part of OTel, when it's just an experiment. A blessed experiment, but an experiment nonetheless.

svrnm replied:
+1 for anticipating the hypothetical problem: my concern is an hypothetical as well: the "on their websites" bothers me, besides API/SDK docs I hope no SIG right now has their own website and everything stays either on github or opentelemetry.io 

> 5. Sandbox projects MUST communicate properly that their project is experimental in any circumstances
> 6. Sandbox projects MUST use resources and infrastructure (repository, slack channel, websites, etc.) provided by the OpenTelemetry community

Maybe that's to harsh, but this is what it boils down to for me.

yurishkuro replied:
"on their websites" can be changed to "in their documentation". A github repo is a form of "website".

CNCF does not prohibit sandbox projects from using CNCF name, they just need to explicitly say "CNCF sandbox" (there are standard logos for those). I agree with @svrnm that "do NOT have the right" wording is odd and in contradiction with the stated benefits of the sandbox program.

jpkrohling replied:
I like the requirement about being an explicit sandbox project. Perhaps this would be better?

```suggestion
5. Sandbox projects MAY mention being part of OpenTelemetry, and in doing so, they MUST explicitly state they are in the sandbox.
```

svrnm replied:
I like that wording!

svrnm commented on text/0233-opentelemetry-sandbox.md:
```diff
@@ -0,0 +1,92 @@
+> Problems the project will solve:
+>
+> Motivation for joining the sandbox:
+>
```
```suggestion
>
> Exit Criteria
>
```

Below you describe possible outcomes for the annual evaluation. i am wondering if the project should be founded with exit in mind, this way it also help existing SIGs to express potential concern (e.g. a SIG might not be able to incorporate without certain criteria met, or a SIG might have an issue with a sandbox forming a new SIG that is relatively close to what they are working on, etc.)

jpkrohling replied:
That's an interesting idea. I'm not 100% sure yet I agree yet, as there might be experiments without a clear idea of where it'll eventually lead to, but I like this idea.

svrnm replied:
I think putting "We don't know yet" or "this is an experiment with no clear outcome yet", maybe this can be optional?

jpkrohling replied:
Sounds good, I added the Exit Criteria there now.

svrnm replied:
TY

jpkrohling commented:
Based on our discussion on the main issue, I'm closing this as I'm no longer comfortable with this proposal.
